File: backend/payment/src/main/java/com/bbmovie/payment/service/zalopay/ZaloHexStringUtil.java
package com.bbmovie.payment.service.zalopay;

import java.nio.charset.StandardCharsets;
import java.util.Locale;

@SuppressWarnings({"squid:S1118", "unused"})
public final class ZaloHexStringUtil {
    private static final byte[] HEX_CHAR_TABLE = {
        (byte) '0', (byte) '1', (byte) '2', (byte) '3',
        (byte) '4', (byte) '5', (byte) '6', (byte) '7',
        (byte) '8', (byte) '9', (byte) 'a', (byte) 'b',
        (byte) 'c', (byte) 'd', (byte) 'e', (byte) 'f'
    };

    public static String byteArrayToHexString(byte[] raw) {
        byte[] hex = new byte[2 * raw.length];
        int index = 0;
        for (byte b : raw) {
            int v = b & 0xFF;
            hex[index++] = HEX_CHAR_TABLE[v >>> 4];
            hex[index++] = HEX_CHAR_TABLE[v & 0x0F];
        }
        return new String(hex, StandardCharsets.UTF_8);
    }

    public static byte[] hexStringToByteArray(String hex) {
        String hexStandard = hex.toLowerCase(Locale.ENGLISH);
        int sz = hexStandard.length() / 2;
        byte[] bytesResult = new byte[sz];
        int idx = 0;

        for (int i = 0; i < sz; i++) {
            byte first = (byte) hexStandard.charAt(idx++);
            byte second = (byte) hexStandard.charAt(idx++);

            first = (first > HEX_CHAR_TABLE[9])
                ? (byte) (first - ((byte) 'a' - 10))
                : (byte) (first - '0');

            second = (second > HEX_CHAR_TABLE[9])
                ? (byte) (second - ((byte) 'a' - 10))
                : (byte) (second - '0');

            bytesResult[i] = (byte) (first * 16 + second);
        }
        return bytesResult;
    }
}

File: backend/payment/src/main/kotlin/com/bbmovie/payment/service/payment/zalopay/ZaloHexStringUtil.kt
package com.bbmovie.payment.service.payment.zalopay

import java.util.Locale

object ZaloHexStringUtil {
    private val HEX_CHAR_TABLE = byteArrayOf(
        '0'.code.toByte(), '1'.code.toByte(), '2'.code.toByte(), '3'.code.toByte(),
        '4'.code.toByte(), '5'.code.toByte(), '6'.code.toByte(), '7'.code.toByte(),
        '8'.code.toByte(), '9'.code.toByte(), 'a'.code.toByte(), 'b'.code.toByte(),
        'c'.code.toByte(), 'd'.code.toByte(), 'e'.code.toByte(), 'f'.code.toByte()
    )

    fun byteArrayToHexString(raw: ByteArray): String {
        val hex = ByteArray(2 * raw.size)
        var index = 0

        for (b in raw) {
            val v = b.toInt() and 0xFF
            hex[index++] = HEX_CHAR_TABLE[v ushr 4]
            hex[index++] = HEX_CHAR_TABLE[v and 0xF]
        }
        return String(hex)
    }

    fun hexStringToByteArray(hex: String): ByteArray {
        val hexStandard = hex.lowercase(Locale.ENGLISH)
        val sz = hexStandard.length / 2
        val bytesResult = ByteArray(sz)

        var idx = 0
        for (i in 0 until sz) {
            bytesResult[i] = hexStandard[idx].code.toByte()
            ++idx
            var tmp: Byte = hexStandard[idx].code.toByte()
            ++idx

            bytesResult[i] = (if (bytesResult[i] > HEX_CHAR_TABLE[9]) {
                (bytesResult[i] - ('a'.code.toByte() - 10))
            } else {
                (bytesResult[i] - '0'.code.toByte())
            }).toByte()
            
            tmp = (if (tmp > HEX_CHAR_TABLE[9]) {
                (tmp - ('a'.code.toByte() - 10))
            } else {
                (tmp - '0'.code.toByte())
            }).toByte()

            bytesResult[i] = (bytesResult[i] * 16 + tmp).toByte()
        }
        return bytesResult
    }
}